vyper/venom/analysis/loop_detection.py

```diff
@@ -0,0 +1,69 @@
+from vyper.utils import OrderedSet
+from vyper.venom.analysis.analysis import IRAnalysis
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.basicblock import IRBasicBlock
+
+
+class LoopDetectionAnalysis(IRAnalysis):
+    """
+    Detects loops and computes basic blocks
+    and the block which is before the loop
+    """
+
+    # key = start of the loop (last bb not in the loop)
+    # value = all the block that loop contains
+    loops: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
+
+    done: OrderedSet[IRBasicBlock]
+    visited: OrderedSet[IRBasicBlock]
+
+    def analyze(self):
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        self.loops: dict[IRBasicBlock, OrderedSet[IRBasicBlock]] = dict()
+        self.done = OrderedSet()
+        self.visited = OrderedSet()
+        entry = self.function.entry
+        self._dfs_r(entry)
+
+    def _dfs_r(self, bb: IRBasicBlock, before: IRBasicBlock = None):
+        if bb in self.visited:
+            assert before is not None, "Loop must have one basic block before it"
+            loop = self._collect_path(before, bb)
+            in_bb = bb.cfg_in.difference({before})
+            assert len(in_bb) == 1, "Loop must have one input basic block"
```

Comment:
- Since we are looking for natural loops, we should probably rename the analysis to `NaturalLoopDetectionAnalysis`
- We should also probably not assert if the loop was not natural, just ignore that loop